lib/ansible/modules/cloud/amazon/aws_glacier_vault.py

```diff
@@ -0,0 +1,227 @@
+
+    def list(self):
+        vaults = self.connection.list_vaults()
+        if "VaultList" not in vaults:
+            self.module.fail_json(msg="Invalid response from Glacier: no VaultList found in response")
+
+        return vaults["VaultList"]
+
+    def exists(self, name):
+        for v in self.list():
+            if "VaultName" not in v:
+                self.module.fail_json(
+                    msg="Invalid response from Glacier: no VaultName found in vault object {0}".format(v))
+
+            if v["VaultName"] == name:
+                return v
+
+        return None
+
+    def create_or_update(self):
+        vault = self.exists(self.name)
+        changed = False
+        if vault is None:
+            vault = self.connection.create_vault(vaultName=self.name)
+            changed = True
+
+        changed |= self.manage_tags()
+
+        # Vault exists, no change to do
+        self.module.exit_json(vault=vault, changed=changed)
+
+    def list_tags(self):
+        tags = self.connection.list_tags_for_vault(vaultName=self.name)
+        if "Tags" not in tags:
+            self.module.fail_json(msg="Invalid response from Glacier: no Tags found in vault object {0}".format(tags))
+        return tags["Tags"]
+
+    def manage_tags(self):
+        changed = False
+        old_tags = self.list_tags()
+        tags_to_set, tags_to_delete = compare_aws_tags(
+            old_tags, self.module.params.get('tags') or {},
+        )
+        if tags_to_set:
+            self.connection.add_tags_to_vault(
+                vaultName=self.name,
+                Tags=tags_to_set)
+            changed |= True
+        if tags_to_delete:
+            self.connection.remove_tags_from_vault(
+                vaultName=self.name,
+                TagKeys=tags_to_delete)
+            changed |= True
+        return changed
+
+    def destroy(self):
+        name = self.module.params.get('name')
+        vault = self.exists(name)
+        if vault is None:
+            self.module.exit_json(changed=False)
+
+        self.connection.delete_vault(vaultName=name)
+        self.module.exit_json(vault=vault, changed=True)
+
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(
+        dict(
+            name=dict(required=True, type='str'),
+            state=dict(default='present', type='str', choices=['present', 'absent']),
+            tags=dict(type='dict'),
+        )
+    )
+
+    module = AnsibleModule(argument_spec=argument_spec)
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto3 required for this module')
+
+    region, ec2_url, aws_connect_params = get_aws_connection_info(module, boto3=True)
+
+    if region:
+        connection = boto3_conn(
+            module,
+            conn_type='client',
+            resource='glacier',
+            region=region,
+            endpoint=ec2_url,
+            **aws_connect_params
+        )
+    else:
+        module.fail_json(msg="region must be specified")
```

Comment:
Let boto3_conn handle region failures - region can be picked up in a number of ways that mean that the `region` variable doesn't get set but the connection is fine.